File: member-service/src/main/java/com/hedvig/memberservice/web/dto/InternalMember.java
package com.hedvig.memberservice.web.dto;

import com.hedvig.memberservice.query.MemberEntity;

import java.time.Instant;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

import com.hedvig.memberservice.util.Gender;
import com.hedvig.memberservice.util.SsnUtilImpl;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.val;

@Data
@NoArgsConstructor
public class InternalMember {
  private Long memberId;

  private String status;

  private String ssn;

  private Gender gender;

  private String firstName;

  private String lastName;

  private String street;

  private Integer floor;

  private String apartment;

  private String city;

  private String zipCode;

  private String country;

  private String email;

  private String phoneNumber;

  private LocalDate birthDate;

  private Instant signedOn;

  private Instant createdOn;

  private String fraudulentStatus;

  private String fraudulentDescription;

  private String acceptLanguage;

  private List<TraceMemberDTO> traceMemberInfo = new ArrayList<>();

  public static InternalMember fromEntity(MemberEntity entity) {
    val dto = new InternalMember();
    dto.setMemberId(entity.getId());
    dto.setStatus(entity.getStatus() != null ? entity.getStatus().name() : "");
    dto.setSsn(entity.getSsn());
    dto.setGender(SsnUtilImpl.getInstance().getGenderFromSsn(entity.getSsn()));
    dto.setFirstName(entity.getFirstName());
    dto.setLastName(entity.getLastName());
    dto.setStreet(entity.getStreet());
    dto.setFloor(entity.getFloor());
    dto.setApartment(entity.getApartment());
    dto.setCity(entity.getCity());
    dto.setZipCode(entity.getZipCode());
    dto.setCountry("SE");
    dto.setEmail(entity.getEmail());
    dto.setPhoneNumber(entity.getPhoneNumber());
    dto.setBirthDate(entity.getBirthDate());
    dto.setSignedOn(entity.getSignedOn());
    dto.setCreatedOn(entity.getCreatedOn());
    dto.setFraudulentStatus(entity.getFraudulentStatus() != null ? entity.getFraudulentStatus().name() : "");
    dto.setFraudulentDescription(entity.getFraudulentDescription());
    dto.setAcceptLanguage(entity.getAcceptLanguage());
    return dto;
  }
}

File: member-service/src/main/java/com/hedvig/memberservice/web/dto/InternalMember.kt
package com.hedvig.memberservice.web.dto

import com.hedvig.memberservice.query.MemberEntity
import com.hedvig.memberservice.util.Gender
import com.hedvig.memberservice.util.SsnUtilImpl.Companion.instance
import java.time.Instant
import java.time.LocalDate
import java.util.ArrayList

class InternalMember(
    val memberId: Long? = null,
    val status: String? = null,
    val ssn: String? = null,
    val gender: Gender? = null,
    val firstName: String? = null,
    val lastName: String? = null,
    val street: String? = null,
    val floor: Int? = null,
    val apartment: String? = null,
    val city: String? = null,
    val zipCode: String? = null,
    val country: String? = null,
    val email: String? = null,
    val phoneNumber: String? = null,
    val birthDate: LocalDate? = null,
    val signedOn: Instant? = null,
    val createdOn: Instant? = null,
    val fraudulentStatus: String? = null,
    val fraudulentDescription: String? = null,
    val acceptLanguage: String? = null,
    val traceMemberInfo: List<TraceMemberDTO> = ArrayList()
) {

    companion object {
        fun fromEntity(entity: MemberEntity) = InternalMember(
            entity.getId(),
            if (entity.getStatus() != null) entity.getStatus().name else "",
            entity.getSsn(),
            instance.getGenderFromSsn(entity.getSsn()),
            entity.getFirstName(),
            entity.getLastName(),
            entity.getStreet(),
            entity.getFloor(),
            entity.getApartment(),
            entity.getCity(),
            entity.zipCode,
            "SE",
            entity.getEmail(),
            entity.getPhoneNumber(),
            entity.getBirthDate(),
            entity.getSignedOn(),
            entity.getCreatedOn(),
            if (entity.getFraudulentStatus() != null) entity.getFraudulentStatus().name else "",
            entity.getFraudulentDescription(),
            entity.getAcceptLanguage()
        )
    }
}
